tools/iotools.go

```diff
@@ -85,3 +86,22 @@ func (r *HashingReader) Read(b []byte) (int, error) {
 
 	return w, err
 }
+
+// RetriableReader wraps a error response of reader as RetriableError()
+type RetriableReader struct {
+	reader io.Reader
+}
+
+func NewRetriableReader(r io.Reader) io.Reader {
+	return &RetriableReader{r}
+}
+
+func (r *RetriableReader) Read(b []byte) (int, error) {
+	n, err := r.reader.Read(b)
+	// EOF is a successful response as it is used to signal a graceful end of input
+	// c.f. https://git.io/v6riQ
+	if err == nil || err == io.EOF {
+		return n, err
+	}
+	return n, errors.NewRetriableError(err)
```

Comment:
Technically this could re-wrap an already retriable error in the case that `r`'s `reader` has another `RetriableReader` underneath, but I don't think it's a huge deal 😄